File: python/hush-icore/hush/core/middleware.py
"""Middleware system for the Hush engine.

Middleware provides hooks into the engine execution lifecycle:
- before_run: modify inputs before execution
- after_run: modify results after execution
- on_error: handle errors during execution
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict

if TYPE_CHECKING:
    from hush.core.ops.graph.graph_op import GraphOp


class Middleware:
    """Base class for engine middleware.

    Subclass and override any of the hooks to add behavior.
    All hooks are async and called in order (before_run) or
    reverse order (after_run, on_error).
    """

    async def before_run(
        self, graph: GraphOp, inputs: Dict[str, Any], context: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Called before graph execution. Can modify inputs.

        Args:
            graph: The GraphOp being executed
            inputs: The input dict (modify and return)
            context: Execution context (user_id, session_id, request_id, etc.)

        Returns:
            The (possibly modified) inputs dict.
        """
        return inputs

    async def after_run(
        self,
        graph: GraphOp,
        inputs: Dict[str, Any],
        result: Dict[str, Any],
        context: Dict[str, Any],
    ) -> Dict[str, Any]:
        """Called after graph execution. Can modify result.

        Args:
            graph: The GraphOp that was executed
            inputs: The original inputs
            result: The result dict (modify and return)
            context: Execution context

        Returns:
            The (possibly modified) result dict.
        """
        return result

    async def on_error(
        self, graph: GraphOp, inputs: Dict[str, Any], error: Exception, context: Dict[str, Any]
    ) -> None:
        """Called when graph execution fails.

        Default behavior re-raises the error. Override to add
        logging, alerting, or error transformation.

        Args:
            graph: The GraphOp that failed
            inputs: The original inputs
            error: The exception that occurred
            context: Execution context
        """
        raise error

File: rust/hush-icore/src/middleware.rs
//! Middleware system for the Hush engine.
//!
//! Middleware provides hooks into the engine execution lifecycle:
//! - `before_run`: modify inputs before execution
//! - `after_run`: modify results after execution
//! - `on_error`: handle errors during execution

use serde_json::Value;

use crate::error::RushError;

/// Trait for engine middleware.
///
/// Implement any subset of hooks to add behavior.
/// `before_run` and `after_run` run in registration order and reverse order respectively.
pub trait Middleware: Send + Sync {
    /// Called before graph execution. Can modify inputs in-place.
    fn before_run(&self, _inputs: &mut Value) -> Result<(), RushError> {
        Ok(())
    }

    /// Called after graph execution. Can modify result in-place.
    fn after_run(&self, _result: &mut Value) -> Result<(), RushError> {
        Ok(())
    }

    /// Called when graph execution fails.
    fn on_error(&self, _error: &RushError) {}
}
